_posts/2025-05-02-stream-windowing-functions.md

```diff
@@ -0,0 +1,383 @@
+        range AS window_start,
+        window_start + INTERVAL 15 MINUTE AS window_end
+    FROM range(
+        '2024-01-01 00:00:00'::TIMESTAMP,
+        '2025-01-01 00:00:00'::TIMESTAMP,
+        INTERVAL 5 MINUTE -- hopping size
+    )
+)
+```
+
+```text
+┌─────────────────────┬─────────────────────┐
+│    window_start     │     window_end      │
+│      timestamp      │      timestamp      │
+├─────────────────────┼─────────────────────┤
+│ 2024-01-01 00:00:00 │ 2024-01-01 00:15:00 │
+│ 2024-01-01 00:05:00 │ 2024-01-01 00:20:00 │
+│ 2024-01-01 00:10:00 │ 2024-01-01 00:25:00 │
+│          ·          │          ·          │
+│          ·          │          ·          │
+│          ·          │          ·          │
+│ 2024-12-31 23:45:00 │ 2025-01-01 00:00:00 │
+│ 2024-12-31 23:50:00 │ 2025-01-01 00:05:00 │
+│ 2024-12-31 23:55:00 │ 2025-01-01 00:10:00 │
+├─────────────────────┴─────────────────────┤
+│ 105408 rows (6 shown)           2 columns │
+└───────────────────────────────────────────┘
+```
+
+We then join the above intervals with the train service data in order to calculate the number of services for each interval:
+
+```sql
+SELECT
+    window_start,
+    window_end,
+    count(*) AS number_of_services
+FROM ams_traffic_v
+INNER JOIN time_range AS ts
+    ON station_service_time BETWEEN ts.window_start AND ts.window_end
+GROUP BY ALL
+ORDER BY 3 DESC, 1 ASC
+LIMIT 5;
+```
+
+resulting in:
+
+```text 
+┌─────────────────────┬─────────────────────┬────────────────────┐
+│    window_start     │     window_end      │ number_of_services │
+│      timestamp      │      timestamp      │       int64        │
+├─────────────────────┼─────────────────────┼────────────────────┤
+│ 2024-02-17 11:25:00 │ 2024-02-17 11:40:00 │                 29 │
+│ 2024-02-17 10:25:00 │ 2024-02-17 10:40:00 │                 28 │
+│ 2024-02-17 12:25:00 │ 2024-02-17 12:40:00 │                 28 │
+│ 2024-02-17 13:25:00 │ 2024-02-17 13:40:00 │                 28 │
+│ 2024-02-17 14:25:00 │ 2024-02-17 14:40:00 │                 28 │
+└─────────────────────┴─────────────────────┴────────────────────┘
+```
+
+Can you imagine how it must have been like in the control room when within 15 minutes, 29 trains were arriving or departing in a station with 15 tracks?
+
+> By applying a `RIGHT OUTER JOIN` in the above query, missing gaps are filled with 0 number of services. Use it with precaution, it is a time consuming query!
+
+## Sliding Windows
+
+[_Sliding windows_](https://learn.microsoft.com/en-us/stream-analytics-query/sliding-window-azure-stream-analytics) are overlapping intervals, but, compared to hopping windows, they are dynamically generated from the time column analyzed, therefore changing when new records are inserted. Sliding windows can be implemented by using the [`RANGE` window framing]({% link docs/stable/sql/functions/window_functions.md %}#framing):
+
+```sql 
+SELECT
+    station_service_time - INTERVAL 15 MINUTE AS window_start, -- window size
```

Comment:
Not feedback, but this is an interesting example of a CSE we might detect! If we could expose the `RANGE` bounds in the SQL dialect, we could avoid duplicating this computation.